src/qengine/state/opencl.cpp

```diff
@@ -10,47 +10,59 @@
 // See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
 // for details.
 
+#include <memory>
+
 #include "oclengine.hpp"
 #include "qengine_opencl.hpp"
 
 namespace Qrack {
 
 #define CMPLX_NORM_LEN 5
 
-void QEngineOCL::InitOCL()
+void QEngineOCL::SetDevice(const int& dID)
 {
-    clObj = OCLEngine::Instance();
+    if (dID >= 0) {
+        deviceID = dID % (clObj->GetDeviceCount());
+    } else {
+        deviceID = -1;
+    }
+    device_context = clObj->GetDeviceContextPtr(deviceID);
+    context = device_context->context;
+    queue = device_context->queue;
```

Comment:
Is a `ReInitOCL` appropriate here?

Reply:
No. This is called in `InitOCL()`.

Reply:
SetDevice is a public API.

Reply:
You're right. When the deviceID is changed, InitOCL() effectively needs to be called again, not ReInitOCL(). So, the entirety of InitOCL() can be moved into SetDeviceID(...), and InitOCL() can call SetDeviceID(...).